### PR TITLE
OpenTelemetry service name should have higher priority than app name when no resource attributes are set

### LINKED ISSUE #33317: OpenTelemetry SDK autoconfiguration ignores OTEL service name in favor of Quarkus app name
### Describe the bug

With https://github.com/quarkusio/quarkus/pull/30033 I expected autoconfiguration work for properties listed here https://github.com/quarkusio/quarkus/blob/main/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/OTelFallbackConfigSourceInterceptor.java#L85 and [OTEL documentation](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#opentelemetry-resource) says that `otel.service.name` should have preference over `otel.resource.attributes`. However, Quarkus OpenTelemetry extension creates Tracer resource directly and is using Quarkus application name (or artifact if not available)
https://github.com/quarkusio/quarkus/blob/43a53588daa126477e24e81c75abfbf017547df5/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerRecorder.java#L52
https://github.com/quarkusio/quarkus/blob/43a53588daa126477e24e81c75abfbf017547df5/extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/tracing/TracerProcessor.java#L158

### Expected behavior

Quarkus documentation says `otel.service.name` should have priority
https://github.com/quarkusio/quarkus/blob/43a53588daa126477e24e81c75abfbf017547df5/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/OTelRuntimeConfig.java#L52

### Actual behavior

Quarkus application name is used.

### How to Reproduce?

Define `otel.service.name` and check traces (I'm using Jaeger). IMHO it's straight forward, but ping me if you really need one.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
Fixes: https://github.com/quarkusio/quarkus/issues/33317

### CODE REVIEW

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerUtil.java [hunk 1]
```diff
@@ -15,16 +15,21 @@ private TracerUtil() {
     }
 
     public static Resource mapResourceAttributes(List<String> resourceAttributes, String serviceName) {
+        // resource service name is always set at build time to the app name, therefore
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerUtil.java:18]: This is getting too hard to read... Would this work?
```
public static Resource mapResourceAttributes(List<String> resourceAttributes, String serviceName) {
    final boolean setServiceName = serviceName != null;
    final AttributesBuilder attributesBuilder = Attributes.builder();

    if (!resourceAttributes.isEmpty()) {
        var attributes = OpenTelemetryUtil.convertKeyValueListToMap(resourceAttributes);
        attributes.forEach(attributesBuilder::put);
    }

    if (setServiceName) {
        attributesBuilder.put(SERVICE_NAME.getKey(), serviceName);
    }

    return Resource.create(attributesBuilder.build());
}
```

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest().setArchiveProducer(
+            () -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(TestSpanExporter.class)
+                    .addClass(TestSpanExporterProvider.class)
+                    .addAsResource(new StringAsset("" +
+                            "quarkus.otel.traces.exporter=test-span-exporter\n" +
+                            "quarkus.otel.bsp.schedule.delay=50\n"), "application.properties")
+                    .addAsResource(
+                            "META-INF/services-config/io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider",
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java:40]: We need to go deeper in this...
We shouldn't need to add this ConfigurableSpanExporterProvider

**Review by brunobat — CHANGES_REQUESTED** [review]

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerUtil.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerUtil.java:18]: sure, will do

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java]: Newly added tests don't use it anymore.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    private static final String SERVICE_NAME = "FrankBullitt";
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest().setArchiveProducer(
+            () -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(TestSpanExporter.class)
+                    .addClass(TestSpanExporterProvider.class)
+                    .addAsResource(new StringAsset("" +
+                            "quarkus.otel.bsp.schedule.delay=50\n"), "application.properties"))
+            .overrideRuntimeConfigKey("quarkus.otel.resource.attributes", "service.name=" + SERVICE_NAME);
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java:38]: We need also a test with properties:
```
quarkus.application.name=app-name
quarkus.otel.service.name=otel-service-name
```
and another one with:
```
quarkus.application.name=app-name
quarkus.otel.service.name=otel-service-name
quarkus.otel.resource.attributes=service.name=otel-resource-service-name
```
To make sure otel-service-name will win.
and another one with:
```
quarkus.application.name=app-name
quarkus.otel.resource.attributes=service.name=otel-resource-service-name
```
To make sure the resource will win.
Maybe creating an abstract test that will just change the properties and maintain the rest of the setup would be more efficient

**Comment by brunobat** [comment]
@michalvavrik I'm creating a set of tests, if you don't mind, I'll push them to your PR

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java:38]: Whether you have application name set or not makes no difference for tests considering resource attribute service name is going to be set anyway.

**Comment by michalvavrik** [comment]
> @michalvavrik I'm creating a set of tests, if you don't mind, I'll push them to your PR

Please go ahead, whatever feels right for you!

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryResourceAttrSvcNameTest.java:38]: > To make sure the resource will win.

No, the test that I added and sets resource service name tests just that. But as said below, please go ahead with changes.

**Comment by brunobat** [comment]
Test io.quarkus.opentelemetry.deployment.OpenTelemetryServiceNameAppNameTest is failing.

**Comment by brunobat** [comment]
> > Test io.quarkus.opentelemetry.deployment.OpenTelemetryServiceNameAppNameTest is failing.
> 
> as it should, it is build time property

It is Runtime since 3.0: https://github.com/quarkusio/quarkus/blob/0d3cb3545c838b7232a32b4a0b7e993c8bd5a496/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/OTelRuntimeConfig.java#L49

**Review by brunobat — APPROVED** [review]
Should be fixed now. Feel free to squash my commits.

**Comment by michalvavrik** [comment]
> Should be fixed now. Feel free to squash my commits.

Done, I also changed `.overrideRuntimeConfigKey("quarkus.application.name", "application-name-must-fail")` to `.overrideConfigKey("quarkus.application.name", "application-name-must-fail")`.

**Review by radcortez — APPROVED** [review]

**Comment by geoand** [comment]
Should https://github.com/quarkusio/quarkus/issues/33317 now we be closed again?

**Comment by michalvavrik** [comment]
> Should #33317 now we be closed again?

Yep

**Pull request merged**